### ISSUE TITLE
Rule proposal: A rule about `filter + map` and `filter + flatMap`

### ISSUE THREAD

**Issue by fisker** [sindresorhus/eslint-plugin-unicorn#1694]
### Description

For the `.filter().flatMap()` case:

I believe it's obviously that `.filter()` is unnecessary.

For the `.filter().map()` case:

If we don't care about the `index` in the callback, in most case `.filter().map()` can be replaced by `.flatMap()`.

Note: The `.flatMap()` can be __SLOWER__ in some cases.

### Fail

```js
array.filter(v => v > 0.5).map(v => v * 2)
```

```js
array.filter(v => v > 0.5).flatMap(v => [v, v * 2])
```

### Pass

```js
array.flatMap(v => v > 0.5 ? [v * 2] : [])
array.flatMap(v => v > 0.5 ? v * 2 : []) // <- Returning value instead of array works too, but I prefer always return an array in `flatMap` 
```

```js
array.flatMap(v => v > 0.5 ? [v, v * 2] : [])
```

**Comment by sindresorhus** [comment]
I personally find the `array.filter(…).map(…)` case more readable, but that might be because I have not used `.flatMap()` a lot yet.

**Comment by fisker** [comment]
I agree the `array.filter(…).map(…)` case more readable sometime. 
My initial idea is about `.filter().flatMap()` too, I thought better bring up  the `map` cases.
Let's focus on `.filter().flatMap()` then.

**Comment by papb** [comment]
To me the first version is slightly, very slightly, more readable (like saves about three seconds of thinking). Probably after a while I'd get used to this `flatMap` trick though, and this difference would go down to zero.

**Comment by fregante** [comment]
Honestly the more I look at this, the more it feels like a hack. This looks like a micro-optimization at the expense of readability.

I do like looping arrays once, but I don't like turning flatMap into a covert filter. However familiar one personally is with `flatMap`, nothing beats `.filter().map()`
